distribution/bwc/build.gradle

```diff
@@ -89,40 +89,69 @@ bwcVersions.forPreviousUnreleased { BwcVersions.UnreleasedVersionInfo unreleased
         commandLine = ['git', 'fetch', '--all']
     }
 
-    String buildMetadataKey = "bwc_refspec_${project.path.substring(1)}"
-    task checkoutBwcBranch(type: LoggedExec) {
-        String refspec = System.getProperty("tests.bwc.refspec.${bwcBranch}", buildMetadata.get(buildMetadataKey, "${remote}/${bwcBranch}"))
-        dependsOn fetchLatest
-        workingDir = checkoutDir
-        commandLine = ['git', 'checkout', refspec]
-        doFirst {
-            println "Checking out elasticsearch ${refspec} for branch ${bwcBranch}"
+    Closure execGit = {   Action<ExecSpec> action ->
+        new ByteArrayOutputStream().withStream { os ->
+            ExecResult result = project.exec { spec ->
+                workingDir = checkoutDir
+                standardOutput os
+                action.execute(spec)
+            }
+            result.assertNormalExitValue()
+            return os.toString().trim()
         }
     }
-
-    File buildMetadataFile = project.file("build/${project.name}/build_metadata")
-    task writeBuildMetadata(type: LoggedExec) {
-        dependsOn checkoutBwcBranch
-        workingDir = checkoutDir
-        commandLine = ['git', 'rev-parse', 'HEAD']
-        ignoreExitValue = true
-        ByteArrayOutputStream output = new ByteArrayOutputStream()
-        standardOutput = output
+    task checkoutBwcBranch() {
+        dependsOn fetchLatest
         doLast {
-            if (execResult.exitValue != 0) {
-                output.toString('UTF-8').eachLine { line -> logger.error(line) }
-                execResult.assertNormalExitValue()
+            String refspec = System.getProperty("tests.bwc.refspec.${bwcBranch}", "${remote}/${bwcBranch}")
+            if (System.getProperty("tests.bwc.checkout.align") != null) {
+                /*
+                    We use a time based approach to make the bwc versions built deterministic and compatible with the current hash.
+                    Most of the time we want to test against latest, but when running delayed exhaustive tests or wanting
+                    reproducible builds we want this to be deterministic by using a hash that was the latest when the current
+                    commit was made.
+
+                    This approach doesn't work with merge commits as these can introduce commits in the chronological order
+                    after the fact e.x. a merge done today can add commits dated with yesterday so the result will no longer be
+                    deterministic.
+
+                    We don't use merge commits, but for additional safety we check that no such commits exist in the time period
```

Comment:
So when a merge is done w/o a merge commit the merged commit dates are the "push" date, yes?

Reply:
Yes, squash merging create a new commit that gets a date and is pushed to the branch immediately if done trough the GH UI. If it's not done from GH UI, `git pull --rebase` creates a new commit that gets a new date and you can't push the old one without pulling from the branch first.  If one doesn't pass `--rebase` a merge commit would be created.